File: apps/api/src/application/queries/events/get-events.query.ts
import { IQuery, IQueryHandler, QueryHandler } from '@nestjs/cqrs';
import { Injectable, ForbiddenException } from '@nestjs/common';
import { SupabaseClient, createClient } from '@supabase/supabase-js';
import { ConfigService } from '@nestjs/config';

export class GetEventsQuery implements IQuery {
  constructor(
    public readonly userId: string,
    public readonly coupleId: string,
    public readonly startDate?: Date,
    public readonly endDate?: Date,
  ) {}
}

export interface EventQueryResult {
  id: string;
  coupleId: string;
  title: string;
  startTime: string;
  endTime: string;
  category: string;
  authorId: string;
  description?: string;
  createdAt: string;
  updatedAt: string;
}

@Injectable()
@QueryHandler(GetEventsQuery)
export class GetEventsQueryHandler implements IQueryHandler<GetEventsQuery> {
  private supabase: SupabaseClient;

  constructor(private readonly configService: ConfigService) {
    this.supabase = createClient(
      this.configService.get<string>('SUPABASE_URL')!,
      this.configService.get<string>('SUPABASE_SERVICE_KEY')!,
    );
  }

  async execute(query: GetEventsQuery): Promise<EventQueryResult[]> {
    // First verify user belongs to couple
    const { data: coupleData } = await this.supabase
      .from('couples')
      .select('user1_id, user2_id')
      .eq('id', query.coupleId)
      .single();

    if (
      !coupleData ||
      (coupleData.user1_id !== query.userId &&
        coupleData.user2_id !== query.userId)
    ) {
      throw new ForbiddenException('User does not belong to this couple');
    }

    let queryBuilder = this.supabase
      .from('events')
      .select('*')
      .eq('couple_id', query.coupleId)
      .order('start_time', { ascending: true });

    if (query.startDate) {
      queryBuilder = queryBuilder.gte(
        'start_time',
        query.startDate.toISOString(),
      );
    }

    if (query.endDate) {
      queryBuilder = queryBuilder.lte('end_time', query.endDate.toISOString());
    }

    const { data, error } = await queryBuilder;

    if (error) {
      throw error;
    }

    return (data || []).map((event) => ({
      id: event.id,
      coupleId: event.couple_id,
      title: event.title,
      startTime: event.start_time,
      endTime: event.end_time,
      category: event.category,
      authorId: event.author_id,
      description: event.description,
      createdAt: event.created_at,
      updatedAt: event.updated_at,
    }));
  }
}

File: apps/api-kotlin/src/main/kotlin/com/couplecalendar/application/query/event/GetEventsQuery.kt
package com.couplecalendar.application.query.event

import com.couplecalendar.application.query.Query
import com.couplecalendar.application.query.QueryHandler
import com.couplecalendar.common.exception.ForbiddenException
import com.couplecalendar.infrastructure.persistence.repository.JpaCoupleRepository
import com.couplecalendar.infrastructure.persistence.repository.JpaEventRepository
import org.springframework.stereotype.Component
import java.time.Instant
import java.util.UUID

data class GetEventsQuery(
    val userId: UUID,
    val coupleId: UUID,
    val startDate: Instant? = null,
    val endDate: Instant? = null
) : Query<List<EventQueryResult>>

data class EventQueryResult(
    val id: String,
    val coupleId: String,
    val title: String,
    val startTime: String,
    val endTime: String,
    val category: String,
    val authorId: String,
    val description: String?,
    val location: String?,
    val createdAt: String,
    val updatedAt: String
)

@Component
class GetEventsQueryHandler(
    private val jpaEventRepository: JpaEventRepository,
    private val jpaCoupleRepository: JpaCoupleRepository
) : QueryHandler<GetEventsQuery, List<EventQueryResult>> {

    override fun handle(query: GetEventsQuery): List<EventQueryResult> {
        val couple = jpaCoupleRepository.findById(query.coupleId).orElse(null)
            ?: return emptyList()

        if (couple.user1Id != query.userId && couple.user2Id != query.userId) {
            throw ForbiddenException("User is not a member of this couple")
        }

        val events = if (query.startDate != null && query.endDate != null) {
            jpaEventRepository.findByCoupleIdAndDateRange(
                query.coupleId,
                query.startDate,
                query.endDate
            )
        } else {
            jpaEventRepository.findByCoupleIdOrderByStartTimeAsc(query.coupleId)
        }

        return events.map { entity ->
            EventQueryResult(
                id = entity.id.toString(),
                coupleId = entity.coupleId.toString(),
                title = entity.title,
                startTime = entity.startTime.toString(),
                endTime = entity.endTime.toString(),
                category = entity.category,
                authorId = entity.authorId.toString(),
                description = entity.memo,
                location = entity.location,
                createdAt = entity.createdAt.toString(),
                updatedAt = entity.updatedAt.toString()
            )
        }
    }
}
